admin/CopyReplication

```diff
@@ -12,35 +12,114 @@ cd "$MB_SERVER_ROOT"
+    fi
+}
+
+create_ftp_subdir() {
+    local subdir="$FTP_DATA_DIR"/"$1"
+    if [ ! -d "$subdir" ]
+    then
+        mkdir -p "$subdir"
+        set_ftp_permissions "$subdir" "$FTP_DIR_MODE"
+    fi
+}
 
 # Replication data, if any, goes to both the backup directory and the
 # FTP directory.
-if [ "$REPLICATION_FILE" ]
-then
-    echo Copying replication packet to backup dir
-    chown "$BACKUP_USER:$BACKUP_GROUP" "$TEMP_DIR"/replication-*.tar.bz2
-    chmod "$BACKUP_FILE_MODE" "$TEMP_DIR"/replication-*.tar.bz2
-    cp -a "$TEMP_DIR"/replication-*.tar.bz2 "$BACKUP_DIR"/replication/
-    if [ "$REPLICATION_FILE_ASC" ]
+for packet in replication-*.tar.bz2
+do
+    packer_number_and_version=$(
+        basename "$packet" .tar.bz2 \
+        | sed -n 's/^replication-\([0-9]\+\(-v2\)\?\)$/\1/p'
+    )
+    packet_number=$(
+        echo "$packer_number_and_version" \
+        | sed -n 's/^\([0-9]\+\)\(-v2\)\?$/\1/p'
+    )
+    packet_version=$(
+        echo "$packer_number_and_version" \
+        | sed -n 's/^[0-9]\+-\(v2\)$/\1/p'
+    )
+
+    if [ -z "$packet_version" ]
     then
-        cp -a "$TEMP_DIR"/replication-*.tar.bz2.asc "$BACKUP_DIR"/replication/
+        packet_version='v1'
     fi
 
-    echo Copying replication packet to FTP dir
-    chown "$FTP_USER:$FTP_GROUP" "$TEMP_DIR"/replication-*.tar.bz2
-    chmod "$FTP_FILE_MODE" "$TEMP_DIR"/replication-*.tar.bz2
-    mv "$TEMP_DIR"/replication-*.tar.bz2 "$FTP_DATA_DIR"/replication/
-    if [ "$REPLICATION_FILE_ASC" ]
+    subdir='replication'
+    create_backup_subdir "$subdir"
+    create_ftp_subdir "$subdir"
+
+    # dbmirror packets go under replication/v1.
+    # dbmirror2 packets go under replication/v2.
+    subdir+="/$packet_version"
+    create_backup_subdir "$subdir"
+    create_ftp_subdir "$subdir"
+
+    # Within v1/v2, packets go into nested subdirs based on their
+    # leading three digits.
+    for (( i=0; i<3; i++ ))
+    do
+        digit="${packet_number:$i:1}"
+        if [ -n "$digit" ]
+        then
+            subdir+="/$digit"
+            create_backup_subdir "$subdir"
+            create_ftp_subdir "$subdir"
+        else
+            break
+        fi
+    done
+
+    echo "Copying $packet to backup dir"
+
+    set_backup_permissions "$packet" "$BACKUP_FILE_MODE"
+    cp -a "$packet" "$BACKUP_DIR"/"$subdir"/
+
+    packet_asc="$packet".asc
+
+    if [ -f "$packet_asc" ]
     then
-        mv "$TEMP_DIR"/replication-*.tar.bz2.asc "$FTP_DATA_DIR"/replication/
+        echo "Copying $packet_asc to backup dir"
+        set_backup_permissions "$packet_asc" "$BACKUP_FILE_MODE"
+        cp -a "$packet_asc" "$BACKUP_DIR"/"$subdir"/
     fi
-fi
 
-# eof
+    echo "Copying $packet to FTP dir"
+
+    set_ftp_permissions "$packet" "$FTP_FILE_MODE"
+    # The packet is copied to replication/ too until the metabrainz.org
+    # code supports fetching packets from the subdirs.
+    cp -a "$packet" "$FTP_DATA_DIR"/replication/
+    cp -a "$packet" "$FTP_DATA_DIR"/"$subdir"/
```

Comment:
To save a bit of space, the second copy could be a hard link to the first copy as both are on the same file system:

```suggestion
    cp -a "$packet" "$FTP_DATA_DIR"/replication/
    ln --physical "$FTP_DATA_DIR"/replication/"$packet" "$FTP_DATA_DIR"/"$subdir"/"$packet"
```